api/v1alpha1/virtualmachinepublishrequest_types.go

```diff
@@ -0,0 +1,530 @@
+//
+// It is recommended to use the MarkComplete function in order to mark
+// a VirtualMachinePublishRequest as complete since the function ensures
+// the status.completionTime field is properly set.
+//
+// This condition is set only when the operation has failed or the
+// VirtualMachineImage resource has been realized.
+func (vmpr VirtualMachinePublishRequest) IsComplete() bool {
+	c := vmpr.getCondition(VirtualMachinePublishRequestConditionComplete)
+	if c != nil && c.Status == metav1.ConditionTrue {
+		return !vmpr.Status.CompletionTime.IsZero()
+	}
+	return false
+}
+
+// IsImageAvailable returns true if there is a status condition of
+// Type=ImageAvailable and Status=True.
+func (vmpr VirtualMachinePublishRequest) IsImageAvailable() bool {
+	c := vmpr.getCondition(VirtualMachinePublishRequestConditionImageAvailable)
+	if c != nil && c.Status == metav1.ConditionTrue {
+		return true
+	}
+	return false
+}
+
+// IsUploaded returns true if there is a status condition of Type=Uploaded
+// and Status=True.
+//
+// This condition is set when the publication has finished uploading the
+// VM to the image registry and is waiting on the VirtualMachineImage
+// resource to be realized.
+func (vmpr VirtualMachinePublishRequest) IsUploaded() bool {
+	c := vmpr.getCondition(VirtualMachinePublishRequestConditionUploaded)
+	if c != nil && c.Status == metav1.ConditionTrue {
+		return true
+	}
+	return false
+}
+
+func (vmpr *VirtualMachinePublishRequest) markCondition(
+	conditionType string,
+	status metav1.ConditionStatus,
+	args ...string,
+) {
+	var (
+		message string
+		reason  string
+	)
+	if len(args) > 0 {
+		reason = args[0]
+	}
+	if len(args) > 1 {
+		message = args[1]
+	}
+	if reason == "" && status == metav1.ConditionTrue {
+		reason = "success"
+	} else {
+		reason = string(status)
+	}
+
+	c := vmpr.getCondition(conditionType)
+	if c != nil && c.Status != status {
+		c.Message = message
+		c.Reason = reason
+		c.Status = status
+		c.LastTransitionTime = metav1.NewTime(time.Now().UTC())
+		return
+	}
+	vmpr.Status.Conditions = append(
+		vmpr.Status.Conditions,
+		metav1.Condition{
+			Type:               conditionType,
+			Message:            message,
+			Reason:             reason,
+			Status:             status,
+			LastTransitionTime: metav1.NewTime(time.Now().UTC()),
+		},
+	)
+}
+
+// MarkComplete adds or updates the status condition Type=Complete
+// sets the condition's Status field to the provided value.
+//
+// This function also updates the status.completionTime field to
+// the current time in UTC. Please note this field is updated every
+// time this function is called. If this function is called with a
+// status of anything other than True, the value of status.completionTIme
+// is zeroed.
+//
+// This function also sets status.ready=true IFF the provided status
+// is True and all other conditions present in the resource have a
+// status=True.
+//
+// The variadic paramter args may be used to set the condition's
```

Comment:
"parameter" looks like it's copy-pasted as it appears in multiple places below too :)